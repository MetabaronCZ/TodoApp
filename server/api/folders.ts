import { Application } from 'express';
import { InferType, object, string } from 'yup';

import { Folder } from 'models/Folder';

import { mockedDb } from '../db';
import { mockApiRequest, mockCreatedId, mockParseParamArray } from '../utils';

const createFolderDataSchema = object({
  title: string().required(),
}).required();

const editFolderDataSchema = object({
  title: string(),
}).required();

export const setApiFolderEndpoints = (app: Application): void => {
  // get folder list
  app.get('/folder', async (req, res) => {
    await mockApiRequest();
    res.json({ data: mockedDb.folders });
  });

  // create folder
  app.post('/folder', async (req, res) => {
    const data = req.body;
    await mockApiRequest();

    try {
      await createFolderDataSchema.validate(data);
      const folderData = data as InferType<typeof createFolderDataSchema>;

      const folder: Folder = {
        ...folderData,
        id: mockCreatedId('FOLDER'),
      };
      mockedDb.folders = [...mockedDb.folders, folder];
      res.json({ data: folder });
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // edit folder
  app.patch('/folder/:id', async (req, res) => {
    const { id } = req.params;
    const data = req.body;
    await mockApiRequest();

    try {
      await editFolderDataSchema.validate(data);
      const folderData = data as InferType<typeof editFolderDataSchema>;

      mockedDb.folders = mockedDb.folders.map((item) => {
        return id === item.id ? { ...item, ...folderData } : item;
      });

      res.json(null);
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // delete folders
  app.delete('/folder', async (req, res) => {
    const { ids } = req.query;
    await mockApiRequest();

    const idsParsed = mockParseParamArray(ids);

    const validIds = mockedDb.folders
      .filter((item) => idsParsed.includes(item.id))
      .map((item) => item.id);

    if (0 === validIds.length) {
      res.status(500).json({ error: 'Could not delete folders: Invalid ID!' });
    } else {
      mockedDb.folders = mockedDb.folders.filter((item) => {
        return !validIds.includes(item.id);
      });

      mockedDb.todos = mockedDb.todos.filter((item) => {
        if (item.folder && validIds.includes(item.folder)) {
          return { ...item, folder: null };
        }
        return item;
      });

      res.json(null);
    }
  });
};
