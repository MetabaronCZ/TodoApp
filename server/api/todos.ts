import { Application } from 'express';
import { InferType, boolean, object, string } from 'yup';

import { Todo } from 'models/Todo';

import { mockedDb } from '../db';
import {
  mockApiRequest,
  mockCreatedId,
  mockParseParamArray,
  mockParseParam,
  mockParseInteger,
} from '../utils';

const createTodoDataSchema = object({
  title: string().required(),
  description: string().required(),
  folder: string().nullable().required(),
  isDone: boolean().required(),
}).required();

const editTodoDataSchema = object({
  title: string(),
  description: string(),
  folder: string().nullable(),
  isDone: boolean(),
}).required();

export const setApiTodoEndpoints = (app: Application): void => {
  // get todo list
  app.get('/todo', async (req, res) => {
    await mockApiRequest();

    // parse params
    const perPage = mockParseInteger(req.query.perPage) ?? 20;
    const page = mockParseInteger(req.query.page) ?? 0;
    const folder = mockParseParam(req.query.folder);
    const query = mockParseParam(req.query.query);
    const sort = mockParseParam(req.query.sort);

    // filter items
    const items = mockedDb.todos.filter((item) => {
      // filter by folder
      if (folder && folder !== item.folder) {
        return false;
      }
      // filter by query string
      if (
        query &&
        !item.title.toLowerCase().includes(query.toLowerCase()) &&
        !item.description.toLowerCase().includes(query.toLowerCase())
      ) {
        return false;
      }
      return true;
    });

    // sort items
    items.sort((a, b) => {
      switch (sort) {
        case 'TITLE_ASC':
          return a.title.localeCompare(b.title);
        case 'TITLE_DESC':
          return b.title.localeCompare(a.title);
        case 'CREATED_ASC':
          return a.created - b.created;
        case 'CREATED_DESC':
          return b.created - a.created;
        case 'DONE_1':
          return a.isDone === b.isDone ? 0 : a.isDone ? -1 : +1;
        case 'DONE_0':
          return a.isDone === b.isDone ? 0 : a.isDone ? +1 : -1;
        default:
          // use CREATED_DESC as default
          return b.created - a.created;
      }
    });

    const start = page * perPage;
    const end = start + perPage;

    res.json({
      data: {
        items: items.slice(start, end),
        count: items.length,
      },
    });
  });

  // get todo detail
  app.get('/todo/:id', async (req, res) => {
    const { id } = req.params;
    await mockApiRequest();

    const todo = mockedDb.todos.find((item) => id === item.id);
    res.json({ data: todo || null });
  });

  // create todo
  app.post('/todo', async (req, res) => {
    const data = req.body;
    await mockApiRequest();

    try {
      await createTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof createTodoDataSchema>;

      const todo: Todo = {
        ...todoData,
        id: mockCreatedId('TODO'),
        created: Date.now(),
      };
      mockedDb.todos = [...mockedDb.todos, todo];
      res.json({ data: todo });
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // edit todo
  app.patch('/todo/:id', async (req, res) => {
    const { id } = req.params;
    const data = req.body;
    await mockApiRequest();

    try {
      await editTodoDataSchema.validate(data);
      const todoData = data as InferType<typeof editTodoDataSchema>;

      mockedDb.todos = mockedDb.todos.map((item) => {
        return id === item.id ? { ...item, ...todoData } : item;
      });

      res.json(null);
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // delete todos
  app.delete('/todo', async (req, res) => {
    const { ids } = req.query;
    await mockApiRequest();

    const idsParsed = mockParseParamArray(ids);

    const validIds = mockedDb.todos
      .filter((item) => idsParsed.includes(item.id))
      .map((item) => item.id);

    if (0 === validIds.length) {
      res.status(500).json({ error: 'Could not delete todos: Invalid ID!' });
    } else {
      mockedDb.todos = mockedDb.todos.filter((item) => {
        return !validIds.includes(item.id);
      });

      res.json(null);
    }
  });
};
