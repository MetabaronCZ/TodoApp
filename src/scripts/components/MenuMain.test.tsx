import React from 'react';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import { Menu } from 'components/common/Menu';
import { MenuMain } from 'components/MenuMain';

import { Folder } from 'models/Folder';
import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';

import { todoFoldersSlice } from 'store/folders';
import { filterTodos } from 'store/todos/actions';

jest.mock('components/common/Menu');

const { setFolders } = todoFoldersSlice.actions;

describe('components/MenuMain', () => {
  it('should render correctly', () => {
    const store = mockStore();

    const { container } = render(
      withMockedProviders(
        <Provider store={store}>
          <MenuMain />
        </Provider>,
      ),
    );
    expect(container).toMatchSnapshot();
  });

  it('should have correct menu items', () => {
    const testData: Folder[] = [
      { id: 'A', title: 'Folder A' },
      { id: 'B', title: 'Folder B' },
    ];
    const mockedMenu = jest.mocked(Menu);
    expect(mockedMenu).toHaveBeenCalledTimes(0);

    const store = mockStore();
    store.dispatch(setFolders(testData));

    const { container } = render(
      withMockedProviders(
        <Provider store={store}>
          <MenuMain />
        </Provider>,
      ),
    );
    expect(container).toMatchSnapshot();
    expect(mockedMenu).toHaveBeenCalledTimes(1);

    const props = mockedMenu.mock.calls[0][0];
    const itemdId = props.items.map((item) => item.id);
    expect(itemdId).toEqual(['', 'A', 'B', 'create']);
    expect(props.items[0].active).toEqual(true); // default selected item
  });

  it('should set filtered folder as active', async () => {
    const testData: Folder[] = [
      { id: 'A', title: 'Folder A' },
      { id: 'B', title: 'Folder B' },
    ];
    const mockedMenu = jest.mocked(Menu);
    expect(mockedMenu).toHaveBeenCalledTimes(0);

    const store = mockStore();
    store.dispatch(setFolders(testData));
    await store.dispatch(filterTodos({ folder: 'B' }));

    render(
      withMockedProviders(
        <Provider store={store}>
          <MenuMain />
        </Provider>,
      ),
    );
    expect(mockedMenu).toHaveBeenCalledTimes(1);

    const props = mockedMenu.mock.calls[0][0];
    const active = props.items.filter((item) => item.active);
    expect(active.length).toEqual(1);
    expect(active[0].id).toEqual('B');
  });
});
