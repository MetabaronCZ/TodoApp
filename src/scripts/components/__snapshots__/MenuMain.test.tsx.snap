// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MenuMain should have correct menu items 1`] = `<div />`;

exports[`components/MenuMain should render correctly 1`] = `<div />`;
